sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/UnitTestHelper.cs

```diff
@@ -0,0 +1,99 @@
+{
+    internal static class UnitTestHelper
+    {
+        private static readonly ISet<KnownType> KnownTestMethodAttributes = new HashSet<KnownType>
+        {
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_TestMethodAttribute,
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_DataTestMethodAttribute,
+            KnownType.NUnit_Framework_TestAttribute,
+            KnownType.NUnit_Framework_TestCaseAttribute,
+            KnownType.NUnit_Framework_TestCaseSourceAttribute,
+            KnownType.NUnit_Framework_TheoryAttribute,
+            KnownType.Xunit_FactAttribute,
+            KnownType.Xunit_TheoryAttribute,
+            KnownType.LegacyXunit_TheoryAttribute
+        };
+
+        private static readonly ISet<KnownType> KnownExpectedExceptionAttributes = new HashSet<KnownType>
+        {
+            // Note: XUnit doesn't have a exception attribute
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_ExpectedExceptionAttribute,
+            KnownType.NUnit_Framework_ExpectedExceptionAttribute
+        };
+
+        private static readonly ISet<KnownType> KnownIgnoreAttributes = new HashSet<KnownType>
+        {
+            // Note: XUnit doesn't have a separate "Ignore" attribute. It has a "Skip" parameter
+            // on the test attribute
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_IgnoreAttribute,
+            KnownType.NUnit_Framework_IgnoreAttribute
+        };
+
+        /// <summary>
+        /// List of partial names that are assumed to indicate an assertion method
+        /// </summary>
+        private static readonly IEnumerable<string> AssertionPartialMethodName =
+            new List<string>
+            {
+                "ASSERT",
+                "SHOULD",
+                "EXPECT",
+                "MUST",
+                "VERIFY",
+                "VALIDATE"
+            };
+        
+        public static bool IsTestMethod(this IMethodSymbol method) =>
+            method.GetAttributes(KnownTestMethodAttributes).Any();
+
+        public static bool HasExpectedExceptionAttribute(this IMethodSymbol method) =>
+            method.GetAttributes().Any(a => a.AttributeClass.IsAny(KnownExpectedExceptionAttributes));
+
+        public static bool IsAssertionMethodName(string name) =>
+            name.SplitCamelCaseToWords().Union(AssertionPartialMethodName).Any();
+
+        public static bool IsMsTestOrNUnitTestIgnored(this IMethodSymbol method)
+        {
+            var attributes = method.GetAttributes();
+            if (attributes.Any(a => a.AttributeClass.IsAny(KnownIgnoreAttributes)))
+            {
+                return true;
+            }
+            return false;
+        }
+
+        public static AttributeData FindXUnitTestAttribute(this IMethodSymbol method) =>
+            method.GetAttributes().FirstOrDefault(a =>
+                    a.AttributeClass.Is(KnownType.Xunit_FactAttribute) |
```

Comment:
This was one of the bugs - the Xunit "ignored test" logic was only looking at _Fact_ rules.